### PR TITLE
Extend Cosmos custom selector to support + when using paths and tags

### LINKED ISSUE #823: Model selection "plus" operator not properly parsed when selecting tags
Migrating an issue raised in the [`#airflow-dbt` slack org](https://apache-airflow.slack.com/archives/C059CC42E9W/p1706214231926849):
image
image

Basically, the "plus" operator isn't properly parsed when used with tags. So `tag:every-hour` works but `+tag:every-hour` or `tag:every-hour+` result in empty graphs.  We use this syntax in our DBT Cloud scheduled jobs, even though it isn't explicitly called out in the [DBT docs](https://docs.getdbt.com/reference/node-selection/graph-operators):
image

I'm happy to try to take a stab at adding this functionality to [selector.py](https://github.com/astronomer/astronomer-cosmos/blob/main/cosmos/dbt/selector.py), unless there's a reason to hold off.

### LINKED ISSUE #1141: [Bug] Unable to create task group from folder when using DBT_MANIFEST
### Astronomer Cosmos Version

Other Astronomer Cosmos version (please specify below)

### If "Other Astronomer Cosmos version" selected, which one?

1.5.1

### dbt-core version

1.8.4

### Versions of dbt adapters

dbt_redshift==1.8.1

### LoadMode

DBT_LS_MANIFEST

### ExecutionMode

LOCAL

### InvocationMode

None

### airflow version

2.9.1

### Operating System

Amazon Linux 2023

### If a you think it's an UI issue, what browsers are you seeing the problem on?

_No response_

### Deployment

Amazon (AWS) MWAA

### Deployment details

I've also tested this in the docker-compose [MWAA local runner](https://github.com/aws/aws-mwaa-local-runner)

### What happened?

When using DBT_MANIFEST as the load_method, cosmos does not generate tasks for models in a path but DBT_LS will.

dbt will allow you to run this as: `dbt ls -s marts.some_folder_with_models` or `dbt ls -s ./models/marts/some_folder_with_models`

## This doesn't work:
```
    project_config=ProjectConfig(
        dbt_project_path=AIRFLOW_HOME / "dags/dbt/dbt-analytics",
        manifest_path=AIRFLOW_HOME / "dags/dbt/dbt-analytics/target/manifest.json",
    ),
    render_config=RenderConfig(
        select=["+marts.some_folder_with_models"],
        load_method=LoadMode.DBT_MANIFEST,
        dbt_executable_path=AIRFLOW_HOME / "dbt_venv/bin/dbt",
    ),
```

image

## This works:
```
    project_config=ProjectConfig(
        dbt_project_path=AIRFLOW_HOME / "dags/dbt/dbt-analytics",
    ),
    render_config=RenderConfig(
        select=["+marts.some_folder_with_models"],
        load_method=LoadMode.DBT_LS,
        dbt_executable_path=AIRFLOW_HOME / "dbt_venv/bin/dbt",
    ),
```

### Relevant log output

_No response_

### How to reproduce

1. in your render config select a folder with models
2. change load_method from DBT_LS to DBT_MANIFEST
3. In the GUI, reload and watch the tasks in the task group disappear

### Anything else :)?

IMO the two methods should return the same DAG, so DBT_MANIFEST should be able to recognize all models and upstream/downstream dependencies in a folder.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Contact Details

mvictoria@healthfirst.org

### PR DESCRIPTION
**Pull request by mvictoria**
## Description

This PR resolves #1141 where dbt projects loaded with DBT_MANIFEST were unable to parse using the GraphSelector.

Changes:
- updated GraphSelector to take into account path and tag dbt selector methods
- update SelectorConfig to use regex to parse selection statement to handle graph and path statements

ToDo:
- ~add tests covering change~
- ~enable config dbt selector method~
- refactor

## Related Issue(s)

closes #1141
closes #823 

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by tatiana — COMMENTED** [review]
Hi @mvictoria, this is looking great, thank you for covering these use-cases that were not previously covered in Cosmos' custom selector.

Please, could you update the description adding also:
Closes: #823

Also, could we rename this PR to something more representative, such as:
"Extend Cosmos custom selector to support + when using paths and tags"

I left some minor comments inline, would love to hear your thoughts on those. Depending on how quickly we iterate on this, we may be able to release it in Cosmos 1.6.0, planned for this week.

File: cosmos/dbt/selector.py [hunk 1]
```diff
@@ -148,22 +150,38 @@ def filter_nodes(self, nodes: dict[str, DbtNode]) -> set[str]:
...
-            node_by_name[node.name] = node_id
+        if PATH_SELECTOR in self.node_name:
+            path_selection = self.node_name[len(PATH_SELECTOR) :]
+            root_nodes.update({node_id for node_id, node in nodes.items() if path_selection in str(node.file_path)})
+
+        elif TAG_SELECTOR in self.node_name:
+            tag_selection = self.node_name[len(TAG_SELECTOR) :]
+            root_nodes.update({node_id for node_id, node in nodes.items() if tag_selection in node.tags})
+
+        elif CONFIG_SELECTOR in self.node_name:
```
Comment (tatiana) [inline cosmos/dbt/selector.py:165]: Please, could we add a test covering this selector - so we make sure it continues working as expected after the refactor?

File: cosmos/dbt/selector.py [hunk 2]
```diff
@@ -210,14 +228,23 @@ def load_from_statement(self, statement: str) -> None:
...
+            if regex_match:
+                precursors, node_name, descendants = regex_match.groups()
+                if node_name is None:
+                    ...
+                elif precursors or descendants:
+                    self._parse_unknown_selector(item)
+                elif node_name.startswith(PATH_SELECTOR):
+                    self._parse_path_selector(item)
+                elif "/" in node_name:
+                    self._parse_path_selector(f"{PATH_SELECTOR}{node_name}")
```
Comment (tatiana) [inline cosmos/dbt/selector.py:241]: Please, could you add a test covering this use-case?

Comment (mvictoria) [reply cosmos/dbt/selector.py:270]: Thanks for the review!  I've added a test for this now here: https://github.com/astronomer/astronomer-cosmos/blob/67cb56ff9c4d7ab20dcd3da926a83aa83bea80b6/tests/dbt/test_selector.py#L194

File: cosmos/dbt/selector.py (hunk 1 above)
Comment (mvictoria) [reply cosmos/dbt/selector.py:169]: added these test cases here: https://github.com/astronomer/astronomer-cosmos/blob/67cb56ff9c4d7ab20dcd3da926a83aa83bea80b6/tests/dbt/test_selector.py#L490-L505

**Comment by mvictoria** [comment]
Hey @tatiana - thanks for the review, I've updated tests for coverage and finished adding ability to use the "config" dbt selector with graph operators.

Please let me know if you have any more feedback!

**Review by tatiana — APPROVED** [review]
Thanks a lot for the contribution, @mvictoria , we'll release this change as part of Cosmos 1.6.

**Pull request merged**